components/script/dom/transformstream.rs

```diff
@@ -0,0 +1,950 @@
+use crate::dom::bindings::codegen::Bindings::TransformerBinding::Transformer;
+use crate::dom::bindings::conversions::ConversionResult;
+use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::reflector::{DomGlobal, Reflector, reflect_dom_object_with_proto};
+use crate::dom::bindings::root::{Dom, DomRoot, MutNullableDom};
+use crate::dom::countqueuingstrategy::{extract_high_water_mark, extract_size_algorithm};
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::readablestream::{ReadableStream, create_readable_stream};
+use crate::dom::types::PromiseNativeHandler;
+use crate::dom::underlyingsourcecontainer::UnderlyingSourceType;
+use crate::dom::writablestream::create_writable_stream;
+use crate::dom::writablestreamdefaultcontroller::UnderlyingSinkType;
+use crate::realms::enter_realm;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+impl js::gc::Rootable for TransformBackPressureChangePromiseFulfillment {}
+
+/// Reacting to backpressureChangePromise as part of
+/// <https://streams.spec.whatwg.org/#transform-stream-default-sink-write-algorithm>
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct TransformBackPressureChangePromiseFulfillment {
+    /// The result of reacting to backpressureChangePromise.
+    #[ignore_malloc_size_of = "Rc is hard"]
+    result_promise: Rc<Promise>,
+
+    #[ignore_malloc_size_of = "mozjs"]
+    chunk: Box<Heap<JSVal>>,
+
+    writable: Dom<WritableStream>,
+
+    controller: Dom<TransformStreamDefaultController>,
+}
+
+impl Callback for TransformBackPressureChangePromiseFulfillment {
+    /// Reacting to backpressureChangePromise with the following fulfillment steps:
+    #[allow(unsafe_code)]
+    fn callback(&self, cx: SafeJSContext, _v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        // Let writable be stream.[[writable]].
+        // Let state be writable.[[state]].
+        // If state is "erroring", throw writable.[[storedError]].
+        if self.writable.is_erroring() {
+            rooted!(in(*cx) let mut error = UndefinedValue());
+            self.writable.get_stored_error(error.handle_mut());
+            self.result_promise.reject(cx, error.handle(), can_gc);
+            return;
+        }
+
+        // Assert: state is "writable".
+        assert!(self.writable.is_writable());
+
+        // Return ! TransformStreamDefaultControllerPerformTransform(controller, chunk).
+        rooted!(in(*cx) let mut chunk = UndefinedValue());
+        chunk.set(self.chunk.get());
+        let result = self
+            .controller
+            .transform_stream_default_controller_perform_transform(
+                cx,
+                &self.writable.global(),
+                chunk.handle(),
+                can_gc,
+            )
+            .expect("perform transform failed");
+        self.result_promise.resolve_native(&result, can_gc);
```

Comment:
I think here you need another native resolve reject handler for `result = Ok(promise)`, where the resolvment should resolve `self.result_promise`, and the rejection should run the  rejection steps of `TransformStreamDefaultControllerPerformTransform`, and also reject `self.result_promise`. (when only rejection or resolving steps are noted, it means the ones not mentioned should pass through, see https://webidl.spec.whatwg.org/#dfn-perform-steps-once-promise-is-settled).

Reply:
Yes that looks correct